### PR TITLE
Avoid top-level `with ...;` in the OCaml packages in `pkgs/development/`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is #208242. This is a pure refactor: there is no functional change contained in this PR.

I really wish to use `callPackage` style dependency injection for these OCaml packages, rather than the style I've used in this PR. I don't know how to do that refactor, however. 

I used the refactor strategy that looks for the uses of names [coming from `lib` and `ocamlPackages`](https://gist.github.com/philiptaron/04326000d5ff20cb72c2805efd09701c). 

<del>Following advice, I also preferred to inherit names from `lib` instead of `builtins` where there was a choice. I also preferred to inherit from `lib` in one place, rather than to have a mixture of `lib.thing` and `inherit (lib) thing`. I can drop those changes if package owners desire.</del> This PR only touches what was in the `with` statements.

There are no rebuilds when I run `nixpkgs-review rev HEAD`, so I've targeted master with this PR.

## Things done

- [x] Tested compilation of all packages that depend on this change using `nixpkgs-review rev HEAD`.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/headache/default.nix [hunk 1]
```diff
@@ -1,6 +1,9 @@
 { lib, fetchFromGitHub, nix-update-script, ocamlPackages }:
 
-with ocamlPackages;
+let
+  inherit (lib) licenses maintainers;
```
Comment (vbgl) [inline pkgs/development/tools/headache/default.nix:4]: Why do you expand the scope of these variables? Declaring variables way before their point of use looks an anti-pattern to me.

File: pkgs/development/tools/ocaml/ocaml-top/default.nix [hunk 2]
```diff
@@ -1,6 +1,11 @@
 { lib, fetchFromGitHub, ocamlPackages }:
 
-with ocamlPackages; buildDunePackage rec {
+let
+  inherit (lib) licenses maintainers;
```
Comment (vbgl) [inline pkgs/development/tools/ocaml/ocaml-top/default.nix:4]: This change is not motivated: there is no “top-level `with`” that is removed there.

**Comment by philiptaron** [comment]
@vbgl: from the PR cover letter

> I also preferred to inherit from `lib` in one place, rather than to have a mixture of `lib.thing` and `inherit (lib) thing`. I can drop those changes if package owners desire.

File: pkgs/development/tools/headache/default.nix (hunk 1 above)
Comment (philiptaron) [reply pkgs/development/tools/headache/default.nix]: I've reverted any changes having to do with `lib`.

File: pkgs/development/tools/ocaml/ocaml-top/default.nix (hunk 2 above)
Comment (philiptaron) [reply pkgs/development/tools/ocaml/ocaml-top/default.nix]: I've reverted any changes having to do with `lib`.

**Comment by philiptaron** [comment]
The only changes in this stack have to do with `with` now. I've updated the PR headmatter.

**Review by vbgl — APPROVED** [review]

**Comment by philiptaron** [comment]
Thanks @vbgl 🙌🏻

**Pull request merged**